### DISCUSSION TITLE
Recreate folder & albums structure

### DISCUSSION THREAD

**Discussion by JyFu-py** [icloud-photos-downloader/icloud_photos_downloader#575] [Q&A]
Hello,

Is there a way to recreate the folder/album structure?

Thanks!

**Comment by AndreyNikiforov** [comment]
What do you want to see/achieve in the end? `/path/to/photos/my_album/2023/01/IMG_1234.HEIC`?

`--album` parameter should allow you to download photos from particular album and `--folder-structure` pattern can generate sub tree from photo creation date.

**Reply by boredazfcuk** [reply]
Quick related question... is it possible download all albums in one run?

Using '--album AllAlbums' or maybe '--album Album1 Album2 "Album 3"' or '--album Album1 --album Album2'

**Reply by AndreyNikiforov** [reply]
I though about that a bit. One approach would be to support patterns in `--album` parameter. Then specifying one album works as before (backward compatibility).

Supporting multiple `-album` params may be good approach too. It is orthogonal to idea with patterns (can work for patterns and for plain album names).

**Comment by JyFu-py** [comment]
I want to reproduce the apple photo directory structure i made on my ipad/iphone, i don't want to lost my albums titles and replace it with only date.

![image](https://user-images.githubusercontent.com/67822884/217482894-1662082a-9ef9-4c9f-8a7e-1cab6164cd80.png)

Thanks!

**Reply by AndreyNikiforov** [reply]
> I want to reproduce the apple photo directory structure i made on my ipad/iphone, i don't want to lost my albums titles and replace it with only date.
> 
> ![image](https://user-images.githubusercontent.com/67822884/217482894-1662082a-9ef9-4c9f-8a7e-1cab6164cd80.png)
> 
> Thanks!

IIUC you created album structure (albums and sub-albums as it seems) and you want all that to be downloaded preserving the structure. And you do not need dates from photo assets to be expanded into subfolders because you did it manually in albums.

I think specifying `--album "2020/20203 Ski UCPA" --folder-structure none` should create necessary hierarchy. Can you try, please? If works, that will take care of one album at a time and ability to sync multiple albums in one run discussed earlier in this thread.

**Reply by JyFu-py** [reply]
not working :'(

```
2023-02-09 00:23:02 DEBUG    Authenticating...
Traceback (most recent call last):
  File "/volume1/@appstore/py3k/usr/local/bin/icloudpd", line 8, in <module>
    sys.exit(main())
  File "/var/packages/py3k/target/usr/local/lib/python3.8/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/var/packages/py3k/target/usr/local/lib/python3.8/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/var/packages/py3k/target/usr/local/lib/python3.8/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/var/packages/py3k/target/usr/local/lib/python3.8/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/var/packages/py3k/target/usr/local/lib/python3.8/site-packages/icloudpd/base.py", line 281, in main
    photos = icloud.photos.albums[album]
KeyError: '2020/202003 Ski UCPA Tignes'
```

i try on album without space and it's not working too

```
~
KeyError: '2018/201801'
```

when i try on folder it seem ok, but i have no items in first level folders

```
>   --album "2020" \
>   --folder-structure none \
>   --threads-num 4 \
>   --set-exif-datetime
2023-02-09 00:23:56 DEBUG    Authenticating...
2023-02-09 00:24:01 DEBUG    Looking up all photos and videos from album 2020...
2023-02-09 00:24:01 INFO     Downloading 0 original photos and videos to /volume1/photos-darkroom/icloudp-test ...
0it [00:00, ?it/s]
2023-02-09 00:24:02 INFO     All photos have been downloaded!
```

Thanks for your help !

**Comment by metahnetoutage** [comment]
I might sound ignorant by what I will say, yet have you found a way to
sync!? The very definition of Syncing is to create an up to date copy of a
state on each device that is Syncing of some storage or space. This way
nothing is lost.

Am I missing something?

**Comment by DmitryDemidov** [comment]
I too would like to be able to download photos in folders corresponding to the names of folders and albums in Photo.app

**Comment by craiganeedham** [comment]
Hi, Ive been working on a PS1 script that : 
1. Checks if Python is installed, if not install it, if so update it
2. Checks if PIP is installed, if not install it, if so update it
3. Installs icloudpb
4. Creates a folder called "iCloud year month day hours mins seconds" on the users desktop
5. Has the iCloud Username and Password hardcoded < for testing while i try and get it working
6. Has a list of Albums to ignore and skip
7. Fetches the names of the albums the user has
8. Sanitizes the album names if they have chars that windows would not like in the name
9. Then the download begins and starts downloading each album one by one, each Album name is placed in the Desktop/iCloud year month day hour min second folder, and the photos and videos are placed in the Album name only (not any other folders).

Its seems to work well... but i have 1 issue... I have the X (twitter app) if i save a photo/video from the X app, it creates an album named "𝕏" - this throws an error up in the code when its parsing the names of the Albums... can anyone help with this, ive tried everything. 

I have attached the ps1 file in a ZIP file:

[iCloud Download.zip](https://github.com/user-attachments/files/19760623/iCloud.Download.zip)

Thanks Craig
